### PR TITLE
graph/path: add doc comments for unexported path construction methods

### PR DESCRIPTION
**Pull request by kortschak**
Please take a look.

### CODE REVIEW

File: graph/path/shortest.go [hunk 1]
```diff
@@ -106,6 +109,8 @@ func (p *Shortest) add(u graph.Node) int {
 	return idx
 }
 
+// set sets the weight of the path from the node in p.nodes indexed by to to the node
+// indexed by to.
```
Comment (romanwerpachowski) [inline graph/path/shortest.go:113]: "to" or "mid"?

File: graph/path/shortest.go [hunk 2]
```diff
@@ -238,6 +249,9 @@ var (
 	defacedBits = math.Float64bits(defaced)
 )
 
+// newAllShortest returns a all-pairs shortest path forrest for paths with the
```
Comment (romanwerpachowski) [inline graph/path/shortest.go:252]: a -> an

Comment (romanwerpachowski) [inline graph/path/shortest.go:252]: "forrest"?

File: graph/path/shortest.go (hunk 1 above)
Comment (kortschak) [reply graph/path/shortest.go:113]: Fixed

File: graph/path/shortest.go (hunk 2 above)
Comment (kortschak) [reply graph/path/shortest.go:252]: Fixed

Comment (kortschak) [reply graph/path/shortest.go:252]: Full of trrees.

Comment (romanwerpachowski) [reply graph/path/shortest.go]: That's a thing in graph theory?

Comment (kortschak) [reply graph/path/shortest.go]: The typo is a Dan thing. I have corrected it.

Comment (romanwerpachowski) [reply graph/path/shortest.go]: Sorry, missed the joke :D

File: graph/path/shortest.go [hunk 3]
```diff
@@ -114,6 +119,12 @@ func (p Shortest) set(to int, weight float64, mid int) {
 		if !ok {
 			p.cycCosts[[2]int{mid, to}] = weight
 		} else if weight < c {
+			// The only ways that we can have a new weight that is
```
Comment (vladimir-ch) [inline graph/path/shortest.go:122]: way

File: graph/path/shortest.go [hunk 4]
```diff
@@ -260,15 +274,21 @@ func newAllShortest(nodes []graph.Node, forward bool) AllShortest {
 	}
 }
 
+// at returns a slice of node indexed into p.nodes for nodes that are mid points
```
Comment (vladimir-ch) [inline graph/path/shortest.go:277]: This sentence is quite hard to parse so not sure but "a slice of nodes" (plural)?

File: graph/path/shortest.go [hunk 5]
```diff
@@ -62,6 +62,9 @@ type Shortest struct {
 	cycCosts map[[2]int]float64
 }
 
+// newShortestFrom returns a shortest path tree for paths from u
+// intialised with the given nodes. The nodes held by the returned
```
[the remark this answers is not in the document]
Comment (kortschak) [reply graph/path/shortest.go:66]: Fixed

File: graph/path/shortest.go (hunk 3 above)
Comment (kortschak) [reply graph/path/shortest.go:122]: No, there are two ways.

File: graph/path/shortest.go (hunk 4 above)
Comment (kortschak) [reply graph/path/shortest.go:277]: A common typo for me is -ed (_p.p._)-> -es (_pl._). I noticed what I though was this here and miscorrected the -es to -ed. I've fixed this and hopefully also clarified it (and below). It now means what I want it to mean, but still is difficult to parse.

**Comment by kortschak** [comment]
PTAL

File: graph/path/shortest.go (hunk 3 above)
Comment (romanwerpachowski) [reply graph/path/shortest.go:124]: I would write "the only two ways" then.

File: graph/path/shortest.go [hunk 6]
```diff
@@ -55,13 +55,18 @@ type Shortest struct {
 	// returned the Shortest value.
 	hasNegativeCycle bool
 
-	// cycCosts holds costs between
-	// pairs of nodes to report negative
-	// cycles. cycCosts is only used
-	// if it is non-nil.
-	cycCosts map[[2]int]float64
+	// negCosts holds negative costs
```
Comment (romanwerpachowski) [inline graph/path/shortest.go:58]: Why are these lines so short?

File: graph/path/shortest.go [hunk 7]
```diff
@@ -264,19 +282,30 @@ func (p *ShortestAlts) add(u graph.Node) int {
...
 	p.next[to] = []int{mid}
 	if weight < 0 {
-		c, ok := p.cycCosts[[2]int{mid, to}]
+		c, ok := p.negCosts[[2]int{mid, to}]
 		if !ok {
-			p.cycCosts[[2]int{mid, to}] = weight
+			p.negCosts[[2]int{mid, to}] = weight
 		} else if weight < c {
-			p.cycCosts[[2]int{mid, to}] = math.Inf(-1)
+			// The only ways that we can have a new weight that is
```
Comment (romanwerpachowski) [inline graph/path/shortest.go:295]: "two ways"

File: graph/path/shortest.go [hunk 8]
```diff
@@ -638,6 +676,10 @@ func (p AllShortest) AllBetween(uid, vid int64) (paths [][]graph.Node, weight fl
 	return paths, weight
 }
 
+// allBetween recursively constructs a slice of paths extending from the node
+// indexed into p.nodes by from to the node indexed by to. The len(seen) must
```
Comment (romanwerpachowski) [inline graph/path/shortest.go:680]: I'm not sure if "The" is needed here.

File: graph/path/shortest.go (hunk 6 above)
Comment (kortschak) [reply graph/path/shortest.go:58]: This is for a couple of loosely related reasons; the first is that it is like this for the same reason that newsprint is in narrow column and the second is that formatting is wysiswyg in source code (that is godoc has no power here). In code, comments that explain hairy sections should make life easier; this is one way to achieve that.

File: graph/path/shortest.go (hunk 7 above)
Comment (kortschak) [reply graph/path/shortest.go:295]: "only" is used here as an emphatic. I think this is the correct wording.

File: graph/path/shortest.go (hunk 8 above)
Comment (kortschak) [reply graph/path/shortest.go:680]: Hmmm. I read this as "The length of seen must ...". However, this isn't user-facing, so I'm not particularly fussed. I'll change it.

File: graph/path/shortest.go (hunk 6 above)
Comment (romanwerpachowski) [reply graph/path/shortest.go:58]: For me it makes the text harder, not easier to read. But I won't insist.

Comment (kortschak) [reply graph/path/shortest.go:58]: Maybe this is a generational thing. I grew up reading newspapers.
